src/main.cpp

```diff
@@ -3349,282 +3363,346 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
-    printf("REORGANIZE: Connect %" PRIszu " blocks; %s..%s\n", vConnect.size(), pfork->GetBlockHash().ToString().substr(0,20).c_str(), pindexNew->GetBlockHash().ToString().substr(0,20).c_str());
+    CBlockIndex* pindexCur = pindexBest;
+    CBlockIndex* pindexNew = mapItem->second;
+    printf("\r\n** Force Reorganize **\r\n");
+    printf(" Current best height %i hash %s\n", pindexCur->nHeight,pindexCur->GetBlockHash().GetHex().c_str());
+    printf(" Target height %i hash %s\n", pindexNew->nHeight,pindexNew->GetBlockHash().GetHex().c_str());
 
-    if (!vDisconnect.empty())
+    CBlock blockNew;
+    if (!blockNew.ReadFromDisk(pindexNew))
     {
-        //Block was disconnected - User is Re-eligibile for staking
-
-        StructCPID sMag = GetInitializedStructCPID2(GlobalCPUMiningCPID.cpid,mvMagnitudes);
+        printf("ForceReorganizeToHash: Fatal Error while reading new best block.\r\n");
+        return false;
+    }
 
-        if (sMag.initialized)
-        {
-            sMag.LastPaymentTime = 0;
-            mvMagnitudes[GlobalCPUMiningCPID.cpid]=sMag;
-        }
-        nLastBlockSolved = 0;
+    //Re-process the last block to trigger orphan and shit
+    if (!SetBestChain(txdb, blockNew, pindexNew))
+    {
+        return error("ForceReorganizeToHash Fatal Error while setting best chain.\r\n");
     }
-    printf("REORGANIZE Disc Size %" PRIszu, vDisconnect.size());
 
-    // Disconnect shorter branch
-    list<CTransaction> vResurrect;
+    AskForOutstandingBlocks(uint256(0));
+    printf("ForceReorganizeToHash: success! height %f hash %s\n\n",(double)pindexBest->nHeight,pindexBest->GetBlockHash().GetHex().c_str());
+    return true;
+}
+
+bool DisconnectBlocksBatch(CTxDB& txdb, list<CTransaction>& vResurrect, unsigned& cnt_dis, CBlockIndex* pcommon)
+{
     set<string> vRereadCPIDs;
-    for( CBlockIndex* pindex : vDisconnect )
+    while(pindexBest != pcommon)
     {
+        if(!pindexBest->pprev)
+            return error("DisconnectBlocksBatch: attempt to reorganize beyond genesis"); /*fatal*/
+
+        if (fDebug) printf("DisconnectBlocksBatch: %s\n",pindexBest->GetBlockHash().GetHex().c_str());
+
         CBlock block;
-        if (!block.ReadFromDisk(pindex))
-            return error("Reorganize() : ReadFromDisk for disconnect failed");
-        if (!block.DisconnectBlock(txdb, pindex))
-            return error("Reorganize() : DisconnectBlock %s failed", pindex->GetBlockHash().ToString().substr(0,20).c_str());
+        if (!block.ReadFromDisk(pindexBest))
+            return error("DisconnectBlocksBatch: ReadFromDisk for disconnect failed"); /*fatal*/
+        if (!block.DisconnectBlock(txdb, pindexBest))
+            return error("DisconnectBlocksBatch: DisconnectBlock %s failed", pindexBest->GetBlockHash().ToString().c_str()); /*fatal*/
+
+        if (!txdb.WriteHashBestChain(pindexBest->GetBlockHash()))
```

Comment:
Should we not do this after setting pindexBest to pindexBest->pprev?

Reply:
that is probably an issue :D
WriteHashBestChain must get the block previous to the disconnected one.